15_3_Receiving_Bitcoind_Notifications_with_C.md

```diff
@@ -0,0 +1,196 @@
+Need to get 496 kB of archives.
+After this operation, 2.148 kB of additional disk space will be used.
+Do you want to continue? [Y/n] y
+Get:1 http://es.archive.ubuntu.com/ubuntu bionic/universe amd64 libczmq4 amd64 4.1.0-2 [145 kB]
+Get:2 http://es.archive.ubuntu.com/ubuntu bionic/universe amd64 libczmq-dev amd64 4.1.0-2 [351 kB]
+Fetched 496 kB in 0s (1.287 kB/s)    
+Selecting previously unselected package libczmq4:amd64.
+(Reading database ... 314055 files and directories currently installed.)
+Preparing to unpack .../libczmq4_4.1.0-2_amd64.deb ...
+Unpacking libczmq4:amd64 (4.1.0-2) ...
+Selecting previously unselected package libczmq-dev:amd64.
+Preparing to unpack .../libczmq-dev_4.1.0-2_amd64.deb ...
+Unpacking libczmq-dev:amd64 (4.1.0-2) ...
+Setting up libczmq4:amd64 (4.1.0-2) ...
+Setting up libczmq-dev:amd64 (4.1.0-2) ...
+Processing triggers for man-db (2.8.3-2ubuntu0.1) ...
+Processing triggers for libc-bin (2.27-3ubuntu1) ...
+```
+
+### 2. Create C Program
+
+Now we've installed ZMQ and we can compile our C program using it's notifications.   
+The program use czmq.h library and receives two parameters as follows,  first param is the point exposed by bitcoind and second the topic about we'll listen.
+
+``` c
+#include <czmq.h>
+int main(int argc, char ** argv) {
+
+  char *zmqserver;
+  char *topic;
+
+  if (argc < 3) {
+    printf("\nUSAGE:\nchainlistener <tcp://localhost:port> <topic>\n\n");
+    return 0;
+  } else {
+    zmqserver = argv[1];
+    topic = argv[2];
+  }
+
+  zsock_t *socket = zsock_new_sub(zmqserver, topic);
+  assert(socket);
+
+  while(1) {
+    zmsg_t *msg;
+    int rc = zsock_recv(socket, "m", &msg);
+    assert(rc == 0);
+
+    char *header = zmsg_popstr(msg);
+    zframe_t *zdata = zmsg_pop(msg);
+    unsigned int *no = (unsigned int*)zmsg_popstr(msg);
+
+    char *data = zframe_strhex(zdata);
+    int len = zframe_size(zdata);
+    printf("Size: %d\n", len);
+    printf("Data: %s", data);
+    printf("\nNo: %d\n", *no);
+
+    free(header);
+    free(data);
+    free(no);
+    free(zdata);
+    zmsg_destroy(&msg);
+    sleep(1);
+ }
+  zsock_destroy(&socket);
+  return 0;
+}
+```
+
+### 3. Compile C program
+
+To compile this C program you have to do it with clang or gcc compiler and test zmq and czmq libraries.
+
+```
+gcc -o chainlistener chainlistener.c -I/usr/local/include -L/usr/local/lib -lzmq -lczmq
+```
+
+### 4. Configure ZMQ on bitcoind
+
+Add this lines to bitcoin.conf file and restart daemon.    
+
+```
+zmqpubrawblock=tcp://127.0.0.1:28332
+zmqpubrawtx=tcp://127.0.0.1:28333
+```
+Then test it's working using this command
+
+```
+$ bitcoin-cli --testnet getzmqnotifications
+```
+Output:
+```
+[
+  {
+    "type": "pubrawblock",
+    "address": "tcp://127.0.0.1:28332",
+    "hwm": 1000
+  },
+  {
+    "type": "pubrawtx",
+    "address": "tcp://127.0.0.1:28333",
+    "hwm": 1000
+  }
+]
+```
+### 5. Execute listener:
+
+```
+$ ./chainlistener tcp://127.0.0.1:28333 rawtx
+Size: 250
```

Comment:
Are there commands that might be be obvious to readers than the rawtx hex? That one's a little intimidating.

Reply:
Take a look,   tried to explain both params,  do we need something more?

Reply:
That's good.